### DISCUSSION TITLE
Multiple step definition files

### DISCUSSION THREAD

**Discussion by jakohans76** [Tyler-Keith-Thompson/CucumberSwift#23] [Q&A] [answered]
Hi again.

In the Ruby and .js Cucumber versions, it is possible to have multiple step definition files to separate by feature and create a better overview as your project becomes bigger. How do I do that with CucumberSwift?

If we use your sample project as an example: 
1) How do I create multiple step definition files based on the `https://github.com/Tyler-Keith-Thompson/CucumberSwiftSample/blob/main/Cocoapods/CucumberSwiftSampleUITests/CucumberSwiftSampleUITests.swift `file?
2) How do I create "CucumberSwiftSampleUITests1.swift" and "CucumberSwiftSampleUITests2.swift" without redeclarations of `SetupSteps` and `Redundant conformance of 'Cucumber' to protocol 'StepImplementation'`.

Thanks in advance (Remember I'm a Swift Newbie but very happy with this library already).

**Comment by Tyler-Keith-Thompson** [comment, marked as answer]
So I normally handle this by creating multiple files and just using functions. Example:

CucumberTests.swift
```swift
extension Cucumber: StepImplementation {
    public var bundle: Bundle {
        class Findme { }
        return Bundle(for: Findme.self)
    }

    public func setupSteps() {
        let app = XCUIApplication()
        BeforeScenario { (_) in
            app.launch()
        }
        AfterScenario { _ in
            app.terminate()
        }
        setupStepsForFeature1(app) //this declared in File 2
    }
}
```

FILE 2:
```swift
import XCTest
extension Cucumber {
    func setupStepsForFeature1(app:XCUIApplication) {
        Given("") { }
        When("") { }
        Then("") { }
    }
}
```

**Reply by jakohans76** [reply]
This is very awesome. Simple and genius. Thanks.

I'm not able to get the setup hooks working. Neither by feature or @tag. Would it be possible for you to give an example since I need a BeforeFeature to work differently for different features. Some I need to log in using the UI and some I need to log in using the launch arguments and bypass the UI.

1) how to do this by a feature name?
```
BeforeFeature { (_) in
    ..
}
```

2) how to do this by @tag?
```
BeforeScenario { (_) in
    ..
}
```

Thanks in advance. I really appreciate you taking the time to help me.

Br,
Jakob

**Reply by Tyler-Keith-Thompson** [reply]
So before I answer a warning: It's easy to get tests that are too tightly coupled to things about the Gherkin that are supposed to be easy to change. If you have a series of things that need to happen before a feature consider using `Background` in Gherkin.

With that out of the way here's a direct answer. You have access to the properties on the `Feature` and `Scenario` object. So in your example:
```swift
BeforeFeature { feature in 
    if feature.tags.contains("thingYouWant") { setupThingYouWantToSetUp() }
    //or
    switch feature.title {
        case "Title of the feature": setupThingYouWantToSetUp()
    }
}
```

The same holds true for the scenario object so
```swift
BeforeScenario { scenario in
    if scenario.tags.contains("thing") { setupThing() }
}
```
